### PR TITLE
Implement #455: auto-install via occ maintenance:install

### LINKED ISSUE #455: Docker image auto-logins first visitor as ADMIN even when admin user + password are set
The docker image at hub.docker.com/_/nextcloud appears to auto-login the first visitor as administrator even when admin user and password are set via `NEXTCLOUD_ADMIN_USER`, `NEXTCLOUD_ADMIN_PASSWORD`. I would suggest this is **extremely unexpected** for an instance with a preconfigured password, and therefore a potential security pitfall for anyone setting up a nextcloud with docker.

### Steps to reproduce
1. Start docker container with fresh/empty data volumes and `NEXTCLOUD_ADMIN_USER`, `NEXTCLOUD_ADMIN_PASSWORD` set, as well as `MYSQL_USER`, `MYSQL_DATABASE`, `MYSQL_PASSWORD, MYSQL_HOST` (such that no setup wizard is required)
2. Wait until instance runs
3. Visit instance

### Expected behaviour
Login mask since an admin password was set, and database and everything else is also preconfigured via environment variables (=no setup wizard necessary)

### Actual behaviour
While there isn't actually a setup wizard, I'm dropped right into the files overview with what appears *administrator rights*. (or anyone else will who manages to visit it before me, I tested this with the help of someone else from their computer) The second visitor gets a login mask, but by then it might be too late

### Server configuration

**Operating system**: Ubuntu

**Web server:** as packaged in the docker image

**Database:** latest redis and mariadb images from dockerhub

**PHP version:** as packaged in the docker image

**Nextcloud version:** (see Nextcloud admin page)

**Updated from an older Nextcloud/ownCloud or fresh install:** no, fresh install

**Where did you install Nextcloud from:** hub.docker.com

**Other things:**

I skipped checking the other things (server config, apps, LDAP, ...) since this happens 100% every time with a fresh docker instance for me. However, if you have trouble reproducing, I'll gladly provide any of the other information if that is needed

### PR DESCRIPTION
**Pull request by SnowMB**
Resolves #455 
Closes #180

For now I tested it locally with the following docker-compose script (on the same level as the repository):

```yaml
version: "3"

services:
  app-sqlite:
    image: nextcloud-test
    build: docker/14.0/apache
    environment:
      - SQLITE_DATABASE=nextcloud
      - NEXTCLOUD_ADMIN_USER=admin
      - NEXTCLOUD_ADMIN_PASSWORD=pw
    ports:
      - "8080:80"

  mysql:
    image: mariadb
    environment:
      - MYSQL_ROOT_PASSWORD=abc
      - MYSQL_DATABASE=nextcloud
      - MYSQL_USER=nextcloud
      - MYSQL_PASSWORD=12345
  app-mysql:
    image: nextcloud-test
    build: docker/14.0/apache
    environment:
      - MYSQL_HOST=mysql
      - MYSQL_DATABASE=nextcloud
      - MYSQL_USER=nextcloud
      - MYSQL_PASSWORD=12345
      - NEXTCLOUD_ADMIN_USER=admin
      - NEXTCLOUD_ADMIN_PASSWORD=pw
    ports:
      - "8081:80"
    depends_on:
      - mysql
      
  postgre:
    image: postgres
    environment: 
      - POSTGRES_PASSWORD=123
      - POSTGRES_DB=lettheinstallcreatethedatabase
      - POSTGRES_USER=nextcloud
  app-postgre:
    image: nextcloud-test
    build: docker/14.0/apache
    environment:
      - POSTGRES_HOST=postgre
      - POSTGRES_PASSWORD=123
      - POSTGRES_DB=nextcloud
      - POSTGRES_USER=nextcloud
      - NEXTCLOUD_ADMIN_USER=admin
      - NEXTCLOUD_ADMIN_PASSWORD=pw
    ports:
      - "8082:80"
    depends_on:
      - postgre
```

I just added a `sleep` for postgre and mysql because otherwise install will fail as long as the database is not finished with it's setup procedure.

### CODE REVIEW

File: 12.0/apache/entrypoint.sh [hunk 1]
```diff
@@ -49,13 +49,50 @@ if version_greater "$image_version" "$installed_version"; then
         fi
     done
 
-    if [ "$installed_version" != "0.0.0.0" ]; then
+    #install
+    if [ "$installed_version" = "0.0.0.0" ]; then
+        echo "New nextcloud instance"
```
[the remark this answers is not in the document]
Comment (tianon) [reply 12.0/apache/entrypoint.sh]: The way we normally deal with that is by only performing this sort of initialization behavior if we're running the actual application (in this case, that'd be either `apache2-*` or `php-fpm`):

https://github.com/docker-library/wordpress/blob/f716ecef37ad5b3954363123aca109b762848b34/docker-entrypoint.sh#L26

> ```bash
> if [[ "$1" == apache2* ]] || [ "$1" == php-fpm ]; then
> 	...
> fi
> ```

[the remark this answers is not in the document]
Comment (SnowMB) [reply 12.0/apache/entrypoint.sh]: I'e looked into the wordpress example and that means basically wrapping everything into the if-block. 
Our entrypoint script uses `#!/bin/sh`, so `[[ "$1" == apache2* ]]` does not work. Any idea how to get around that?

File: 12.0/apache/entrypoint.sh [hunk 2]
```diff
@@ -49,13 +49,50 @@ if version_greater "$image_version" "$installed_version"; then
...
+        if [ ! -z ${NEXTCLOUD_ADMIN_USER+x} ] && [ ! -z ${NEXTCLOUD_ADMIN_PASSWORD+x} ]; then
+            install_options="-n --admin-user \"$NEXTCLOUD_ADMIN_USER\" --admin-pass \"$NEXTCLOUD_ADMIN_PASSWORD\""
+            if [ ! -z ${NEXTCLOUD_TABLE_PREFIX+x} ]; then
+                install_options="$install_options --database-table-prefix \"$NEXTCLOUD_TABLE_PREFIX\""
+            fi
+            if [ ! -z ${NEXTCLOUD_DATA_DIR+x} ]; then
+                install_options="$install_options --data-dir \"$NEXTCLOUD_DATA_DIR\""
+            fi
+            
+            if [ ! -z ${SQLITE_DATABASE+x} ]; then
```
[the remark this answers is not in the document]
Comment (SnowMB) [reply 12.0/apache/entrypoint.sh]: done

File: 12.0/apache/entrypoint.sh [hunk 3]
```diff
@@ -49,13 +49,50 @@ if version_greater "$image_version" "$installed_version"; then
         fi
     done
 
-    if [ "$installed_version" != "0.0.0.0" ]; then
+    #install
+    if [ "$installed_version" = "0.0.0.0" ]; then
+        echo "New nextcloud instance"
+        
+        if [ ! -z ${NEXTCLOUD_ADMIN_USER+x} ] && [ ! -z ${NEXTCLOUD_ADMIN_PASSWORD+x} ]; then
```
[the remark this answers is not in the document]
Comment (SnowMB) [reply 12.0/apache/entrypoint.sh]: done

File: 12.0/apache/entrypoint.sh [hunk 4]
```diff
@@ -49,13 +49,50 @@ if version_greater "$image_version" "$installed_version"; then
...
+            install_options="-n --admin-user \"$NEXTCLOUD_ADMIN_USER\" --admin-pass \"$NEXTCLOUD_ADMIN_PASSWORD\""
+            if [ ! -z ${NEXTCLOUD_TABLE_PREFIX+x} ]; then
+                install_options="$install_options --database-table-prefix \"$NEXTCLOUD_TABLE_PREFIX\""
```
[the remark this answers is not in the document]
Comment (SnowMB) [reply 12.0/apache/entrypoint.sh]: done

**Comment by tilosp** [comment]
i think we can remove a few lines from the `autoconfig.php` file
https://github.com/nextcloud/docker/blob/69e982d7c7d7d00e04ce79a353fc275a9c4496e9/.config/autoconfig.php#L29-L33

File: 12.0/apache/entrypoint.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (tianon) [reply 12.0/apache/entrypoint.sh]: For pure-POSIX you could use something like `[ "${1#apache2}" != "$1" ]` (strip off any "apache2" prefix and see if the resulting string has changed).

File: 12.0/apache/entrypoint.sh [hunk 5]
```diff
@@ -19,43 +19,84 @@ run_as() {
     fi
 }
 
-installed_version="0.0.0.0"
-if [ -f /var/www/html/version.php ]; then
+if expr "$1" : "apache*" 1>/dev/null || [ "$1" = "php-fpm" ]; then
```
[the remark this answers is not in the document]
Comment (SnowMB) [reply 12.0/apache/entrypoint.sh]: You're right. The correct expression would have been "apache.*". With the above i'm looking for 0 or more 'e's. stupid me :grinning: 

Your solution works just fine. So I used this one.

File: 12.0/apache/entrypoint.sh [hunk 6]
```diff
@@ -19,43 +19,84 @@ run_as() {
...
-    if [ "$installed_version" != "0.0.0.0" ]; then
-        run_as 'php /var/www/html/occ upgrade'
+        #install
+        if [ "$installed_version" = "0.0.0.0" ]; then
+            echo "New nextcloud instance"
+            
+            if [ -n "${NEXTCLOUD_ADMIN_USER+x}" ] && [ -n "${NEXTCLOUD_ADMIN_PASSWORD+x}" ]; then
+                install_options="-n --admin-user \"$NEXTCLOUD_ADMIN_USER\" --admin-pass \"$NEXTCLOUD_ADMIN_PASSWORD\""
+                if [ -n "${NEXTCLOUD_TABLE_PREFIX+x}" ]; then
+                    install_options="$install_options --database-table-prefix \"$NEXTCLOUD_TABLE_PREFIX\""
```
Comment (tilosp) [inline 12.0/apache/entrypoint.sh]: this will brake if some var conatains a `"` or `\`

Comment (tilosp) [reply 12.0/apache/entrypoint.sh]: Only possible solution would be to let the shell in `run_as` fill in the vars.
like this:
```bash
install_options='-n --admin-user "$NEXTCLOUD_ADMIN_USER" --admin-pass "$NEXTCLOUD_ADMIN_PASSWORD"'
```

Comment (tilosp) [reply 12.0/apache/entrypoint.sh]: and `su` should preserve the environment.
```diff
run_as() {
    if [ "$(id -u)" = 0 ]; then
-       su - www-data -s /bin/sh -c "$1"
+       su -p www-data -s /bin/sh -c "$1"
    else
        sh -c "$1"
    fi
}

Comment (SnowMB) [reply 12.0/apache/entrypoint.sh]: Done, sorry for the delay. I'm quite busy at the moment :smile:

**Comment by tilosp** [comment]
@SnowMB thanks a lot for this great pr

**Pull request merged**